LLVM 19: Build failures

The following needs to be ported to the latest LLVM tip:

```
In file included from /tmp/cvise/clang_delta/ReplaceDerivedClass.cpp:19:
/tmp/cvise/clang_delta/CommonRenameClassRewriteVisitor.h: In member function 'bool clang_delta_common_visitor::CommonRenameClassRewriteVisitor<T>::VisitClassTemplatePartialSpecializationDecl(clang::ClassTemplatePartialSpecializationDecl*)':
/tmp/cvise/clang_delta/CommonRenameClassRewriteVisitor.h:147:39: error: 'class clang::ClassTemplatePartialSpecializationDecl' has no member named 'getTypeAsWritten'
  147 |     const TypeSourceInfo *TyInfo = D->getTypeAsWritten();
      |                                       ^~~~~~~~~~~~~~~~
/tmp/cvise/clang_delta/RemoveNamespace.cpp: In member function 'bool RemoveNamespaceRewriteVisitor::VisitClassTemplatePartialSpecializationDecl(clang::ClassTemplatePartialSpecializationDecl*)':
/tmp/cvise/clang_delta/RemoveNamespace.cpp:442:39: error: 'class clang::ClassTemplatePartialSpecializationDecl' has no member named 'getTypeAsWritten'
  442 |     const TypeSourceInfo *TyInfo = D->getTypeAsWritten();
      |                                       ^~~~~~~~~~~~~~~~
```

Can @nickdesaulniers or @strimo378 help me, please?

Thanks for the suggestions. With the following changes applied:

```patch
diff --git a/clang_delta/CommonRenameClassRewriteVisitor.h b/clang_delta/CommonRenameClassRewriteVisitor.h
index f2daeb2..ecbe4f1 100644
--- a/clang_delta/CommonRenameClassRewriteVisitor.h
+++ b/clang_delta/CommonRenameClassRewriteVisitor.h
@@ -144,7 +144,7 @@ bool CommonRenameClassRewriteVisitor<T>::
 
   std::string Name;
   if (getNewName(CXXRD, Name)) {
-    const TypeSourceInfo *TyInfo = D->getTypeAsWritten();
+    const TypeSourceInfo *TyInfo = D->getTemplateArgsAsWritten()->getTemplateArgs()->getTypeSourceInfo();
     if (!TyInfo)
       return true;
     TypeLoc TyLoc = TyInfo->getTypeLoc();
diff --git a/clang_delta/RemoveNamespace.cpp b/clang_delta/RemoveNamespace.cpp
index b31c00e..8650c49 100644
--- a/clang_delta/RemoveNamespace.cpp
+++ b/clang_delta/RemoveNamespace.cpp
@@ -439,7 +439,7 @@ bool RemoveNamespaceRewriteVisitor::VisitClassTemplatePartialSpecializationDecl(
 
   std::string Name;
   if (ConsumerInstance->getNewName(CXXRD, Name)) {
-    const TypeSourceInfo *TyInfo = D->getTypeAsWritten();
+    const TypeSourceInfo *TyInfo = D->getTemplateArgsAsWritten()->getTemplateArgs()->getTypeSourceInfo();
     if (!TyInfo)
       return true;
     TypeLoc TyLoc = TyInfo->getTypeLoc();
```

I've got the following failing tests:

```
cls = <class 'test_clang_delta.TestClangDelta'>, testcase = 'rename-class/instantiation.cpp', arguments = '--transformation=rename-class --counter=1', output_file = 'rename-class/instantiation.output'
E       AssertionError: assert 'template <typename T>\nstruct A {};\n\ntemplate struct AAA<int>;\n' == 'template <typename T>\nstruct A {};\n\ntemplate struct A<int>;\n'
E           template <typename T>
E           struct A {};
E           
E         - template struct A<int>;
E         + template struct AAA<int>;
E         ?                 ++
```

Here it seems the renaming happens only for the original struct, not the later one.

```
cls = <class 'test_clang_delta.TestClangDelta'>, testcase = 'rename-class/partial_specialization.cpp', arguments = '--transformation=rename-class --counter=1', output_file = 'rename-class/partial_specialization.output'
E       AssertionError: assert 'template <typename T, int N>\nstruct A {\n  T value() const { return N; }\n};\n\ntemplate <typename T>\nstruct S1 <A 3> {\n  T value() const { return 0; }\n};\n\n' == 'template <typename T, int N>\nstruct A {\n  T value() const { return N; }\n};\n\ntemplate <typename T>\nstruct A <T, 3> {\n  T value() const { return 0; }\n};\n\n'
E           template <typename T, int N>
E           struct A {
E             T value() const { return N; }
E           };
E           
E           template <typename T>
E         - struct A <T, 3> {
E         ?         ----
E         + struct S1 <A 3> {
E         ?        ++++
E             T value() const { return 0; }
E           };
```

And here it seems the location really points to the first template argument (rather than to the `S1`).

```
cls = <class 'test_clang_delta.TestClangDelta'>, testcase = 'rename-class/specialization.cpp', arguments = '--transformation=rename-class --counter=1', output_file = 'rename-class/specialization.output'
E       AssertionError: assert 'template <typename T>\nstruct A {};\n\ntemplate<> struct AAA<int> {};\n' == 'template <typename T>\nstruct A {};\n\ntemplate<> struct A<int> {};\n'
E           template <typename T>
E           struct A {};
E           
E         - template<> struct A<int> {};
E         + template<> struct AAA<int> {};
E         ?                   ++
```

This one is similar to the first failing test-case.

--- a/clang_delta/CommonRenameClassRewriteVisitor.h
+++ b/clang_delta/CommonRenameClassRewriteVisitor.h
@@ -144,11 +144,7 @@ bool CommonRenameClassRewriteVisitor<T>::
 
   std::string Name;
   if (getNewName(CXXRD, Name)) {
-    const TypeSourceInfo *TyInfo = D->getTypeAsWritten();
-    if (!TyInfo)
-      return true;
-    TypeLoc TyLoc = TyInfo->getTypeLoc();
-    SourceLocation LocStart = TyLoc.getBeginLoc();
+    SourceLocation LocStart = D->getLocation();
     TransAssert(LocStart.isValid() && "Invalid Location!");
     TheRewriter->ReplaceText(LocStart, CXXRD->getNameAsString().size(), Name);
   }

--- a/clang_delta/CommonRenameClassRewriteVisitor.h
+++ b/clang_delta/CommonRenameClassRewriteVisitor.h
@@ -57,9 +57,6 @@ public:
   bool VisitDependentTemplateSpecializationTypeLoc(
          DependentTemplateSpecializationTypeLoc DTSLoc);
 
-  bool VisitClassTemplatePartialSpecializationDecl(
-         ClassTemplatePartialSpecializationDecl *D);
-
   bool VisitClassTemplateSpecializationDecl(
          ClassTemplateSpecializationDecl *TSD);
 
@@ -122,35 +119,6 @@ bool CommonRenameClassRewriteVisitor<T>::TraverseConstructorInitializer(
   return true;
 }
 
-template<typename T> 
-bool CommonRenameClassRewriteVisitor<T>::
-     VisitClassTemplatePartialSpecializationDecl(
-       ClassTemplatePartialSpecializationDecl *D)
-{
-  const Type *Ty = D->getInjectedSpecializationType().getTypePtr();
-  TransAssert(Ty && "Bad TypePtr!");
-  const TemplateSpecializationType *TST = 
-    dyn_cast<TemplateSpecializationType>(Ty);
-  TransAssert(TST && "Bad TemplateSpecializationType!");
-
-  TemplateName TplName = TST->getTemplateName();
-  const TemplateDecl *TplD = TplName.getAsTemplateDecl();
-  TransAssert(TplD && "Invalid TemplateDecl!");
-  NamedDecl *ND = TplD->getTemplatedDecl();
-  TransAssert(ND && "Invalid NamedDecl!");
-
-  const CXXRecordDecl *CXXRD = dyn_cast<CXXRecordDecl>(ND);
-  TransAssert(CXXRD && "Invalid CXXRecordDecl!");
-
-  std::string Name;
-  if (getNewName(CXXRD, Name)) {
-    SourceLocation LocStart = D->getLocation();
-    TransAssert(LocStart.isValid() && "Invalid Location!");
-    TheRewriter->ReplaceText(LocStart, CXXRD->getNameAsString().size(), Name);
-  }
-  return true;
-}
-
 // ISSUE: I am not sure why, but RecursiveASTVisitor doesn't recursively
 // visit base classes from explicit template specialization, e.g.,
 //   struct A { };
@@ -162,14 +130,24 @@ template<typename T>
 bool CommonRenameClassRewriteVisitor<T>::VisitClassTemplateSpecializationDecl(
        ClassTemplateSpecializationDecl *TSD)
 {
-  if (!TSD->isExplicitSpecialization() || !TSD->isCompleteDefinition())
+  if (!TSD->isExplicitInstantiationOrSpecialization())
     return true;
 
-  for (CXXRecordDecl::base_class_const_iterator I = TSD->bases_begin(),
-       E = TSD->bases_end(); I != E; ++I) {
-    TypeSourceInfo *TSI = (*I).getTypeSourceInfo();
-    TransAssert(TSI && "Bad TypeSourceInfo!");
-    getDerived().TraverseTypeLoc(TSI->getTypeLoc());
+  const CXXRecordDecl *CXXRD = TSD->getSpecializedTemplate()->getTemplatedDecl();
+  std::string Name;
+  if (getNewName(CXXRD, Name)) {
+    SourceLocation LocStart = TSD->getLocation();
+    TransAssert(LocStart.isValid() && "Invalid Location!");
+    TheRewriter->ReplaceText(LocStart, CXXRD->getNameAsString().size(), Name);
+  }
+
+  if (TSD->isExplicitSpecialization() && TSD->isCompleteDefinition()) {
+    for (CXXRecordDecl::base_class_const_iterator I = TSD->bases_begin(),
+         E = TSD->bases_end(); I != E; ++I) {
+      TypeSourceInfo *TSI = (*I).getTypeSourceInfo();
+      TransAssert(TSI && "Bad TypeSourceInfo!");
+      getDerived().TraverseTypeLoc(TSI->getTypeLoc());
+    }
   }
   return true;
 }

--- a/clang_delta/RemoveNamespace.cpp
+++ b/clang_delta/RemoveNamespace.cpp
@@ -439,11 +439,7 @@ bool RemoveNamespaceRewriteVisitor::VisitClassTemplatePartialSpecializationDecl(
 
   std::string Name;
   if (ConsumerInstance->getNewName(CXXRD, Name)) {
-    const TypeSourceInfo *TyInfo = D->getTypeAsWritten();
-    if (!TyInfo)
-      return true;
-    TypeLoc TyLoc = TyInfo->getTypeLoc();
-    SourceLocation LocStart = TyLoc.getBeginLoc();
+    SourceLocation LocStart = D->getLocation();
     TransAssert(LocStart.isValid() && "Invalid Location!");
     ConsumerInstance->TheRewriter.ReplaceText(
       LocStart, CXXRD->getNameAsString().size(), Name);

--- a/clang_delta/RemoveNamespace.cpp
+++ b/clang_delta/RemoveNamespace.cpp
@@ -91,9 +91,6 @@ public:
   bool VisitTemplateSpecializationTypeLoc(
          TemplateSpecializationTypeLoc TSPLoc);
 
-  bool VisitClassTemplatePartialSpecializationDecl(
-         ClassTemplatePartialSpecializationDecl *D);
-
   bool VisitDependentTemplateSpecializationTypeLoc(
          DependentTemplateSpecializationTypeLoc DTSLoc);
 
@@ -132,14 +129,25 @@ bool RemoveNamespaceASTVisitor::VisitNamespaceDecl(NamespaceDecl *ND)
 bool RemoveNamespaceRewriteVisitor::VisitClassTemplateSpecializationDecl(
        ClassTemplateSpecializationDecl *TSD)
 {
-  if (!TSD->isExplicitSpecialization() || !TSD->isCompleteDefinition())
+  if (!TSD->isExplicitInstantiationOrSpecialization())
     return true;
 
-  for (CXXRecordDecl::base_class_const_iterator I = TSD->bases_begin(),
-       E = TSD->bases_end(); I != E; ++I) {
-    TypeSourceInfo *TSI = (*I).getTypeSourceInfo();
-    TransAssert(TSI && "Bad TypeSourceInfo!");
-    TraverseTypeLoc(TSI->getTypeLoc());
+  const CXXRecordDecl *CXXRD = TSD->getSpecializedTemplate()->getTemplatedDecl();
+  std::string Name;
+  if (ConsumerInstance->getNewName(CXXRD, Name)) {
+    SourceLocation LocStart = TSD->getLocation();
+    TransAssert(LocStart.isValid() && "Invalid Location!");
+    ConsumerInstance->TheRewriter.ReplaceText(
+      LocStart, CXXRD->getNameAsString().size(), Name);
+  }
+
+  if (TSD->isExplicitSpecialization() && TSD->isCompleteDefinition()) {
+    for (CXXRecordDecl::base_class_const_iterator I = TSD->bases_begin(),
+         E = TSD->bases_end(); I != E; ++I) {
+      TypeSourceInfo *TSI = (*I).getTypeSourceInfo();
+      TransAssert(TSI && "Bad TypeSourceInfo!");
+      TraverseTypeLoc(TSI->getTypeLoc());
+    }
   }
   return true;
 }
@@ -416,34 +424,6 @@ bool RemoveNamespaceRewriteVisitor::VisitTemplateSpecializationTypeLoc(
       LocStart, CXXRD->getNameAsString().size(), Name);
   }
 
-  return true;
-}
-
-bool RemoveNamespaceRewriteVisitor::VisitClassTemplatePartialSpecializationDecl(
-       ClassTemplatePartialSpecializationDecl *D)
-{
-  const Type *Ty = D->getInjectedSpecializationType().getTypePtr();
-  TransAssert(Ty && "Bad TypePtr!");
-  const TemplateSpecializationType *TST =
-    dyn_cast<TemplateSpecializationType>(Ty);
-  TransAssert(TST && "Bad TemplateSpecializationType!");
-
-  TemplateName TplName = TST->getTemplateName();
-  const TemplateDecl *TplD = TplName.getAsTemplateDecl();
-  TransAssert(TplD && "Invalid TemplateDecl!");
-  NamedDecl *ND = TplD->getTemplatedDecl();
-  TransAssert(ND && "Invalid NamedDecl!");
-
-  const CXXRecordDecl *CXXRD = dyn_cast<CXXRecordDecl>(ND);
-  TransAssert(CXXRD && "Invalid CXXRecordDecl!");
-
-  std::string Name;
-  if (ConsumerInstance->getNewName(CXXRD, Name)) {
-    SourceLocation LocStart = D->getLocation();
-    TransAssert(LocStart.isValid() && "Invalid Location!");
-    ConsumerInstance->TheRewriter.ReplaceText(
-      LocStart, CXXRD->getNameAsString().size(), Name);
-  }
   return true;
 }
 

